libretroshare/src/retroshare/rsids.h

```diff
@@ -19,161 +20,219 @@
-				ret.bytes[i] = ~bytes[i];
-			return ret;
-		}
-		inline t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE,UNIQUE_IDENTIFIER>
-		operator| (const t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE,UNIQUE_IDENTIFIER>& fp) const
-		{
-			t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE,UNIQUE_IDENTIFIER> ret;
-			for(uint32_t i=0; i < ID_SIZE_IN_BYTES; ++i)
-				ret.bytes[i] = bytes[i] | fp.bytes[i];
-			return ret;
-		}
+	inline bool operator==(const Id_t& fp) const
+	{ return !memcmp(bytes, fp.bytes, ID_SIZE_IN_BYTES); }
+
+	inline bool operator!=(const Id_t& fp) const
+	{ return !!memcmp(bytes, fp.bytes, ID_SIZE_IN_BYTES); }
+
+	inline bool operator< (const Id_t& fp) const
+	{ return (memcmp(bytes, fp.bytes, ID_SIZE_IN_BYTES) < 0); }
+
+	inline Id_t operator~ () const
+	{
+		Id_t ret;
+		for(uint32_t i=0; i < ID_SIZE_IN_BYTES; ++i)
+			ret.bytes[i] = ~bytes[i];
+		return ret;
+	}
+
+	inline Id_t operator| (const Id_t& fp) const
+	{
+		Id_t ret;
+		for(uint32_t i=0; i < ID_SIZE_IN_BYTES; ++i)
+			ret.bytes[i] = bytes[i] | fp.bytes[i];
+		return ret;
+	}
 
-		inline bool isNull() const 
-		{ 
-			for(uint32_t i=0;i<SIZE_IN_BYTES;++i) 
-				if(bytes[i] != 0)
-					return false ;
-			return true ;
-		} 
+	inline bool isNull() const
+	{
+		for(uint32_t i=0; i < SIZE_IN_BYTES; ++i)
+		if(bytes[i] != 0) return false;
+		return true;
+	}
 
-		friend std::ostream& operator<<(std::ostream& out,const t_RsGenericIdType<ID_SIZE_IN_BYTES,UPPER_CASE,UNIQUE_IDENTIFIER>& id)
+	friend std::ostream& operator<<(std::ostream& out, const Id_t& id)
+	{
+		switch (UNIQUE_IDENTIFIER)
+		{
+		case RsGenericIdType::PGP_FINGERPRINT:
 		{
-			return out << id.toStdString(UPPER_CASE) ;
+			uint8_t index = 0;
+			for(char c : id.toStdString())
+			{
+				out << c;
+				if(++index % 4 == 0 && index < id.SIZE_IN_BYTES*2) out << ' ';
+			}
+		}
+		break;
+		default: out << id.toStdString(UPPER_CASE); break;
 		}
 
-		inline std::string toStdString() const { return toStdString(UPPER_CASE) ; }
+		return out;
+	}
 
-        inline static uint32_t serial_size() { return SIZE_IN_BYTES ; }
-        bool serialise(void *data,uint32_t pktsize,uint32_t& offset) const
-		{
-			if(offset + SIZE_IN_BYTES > pktsize)
-				return false ;
+	inline std::string toStdString() const { return toStdString(UPPER_CASE); }
 
-			memcpy(&((uint8_t*)data)[offset],bytes,SIZE_IN_BYTES) ;
-			offset += SIZE_IN_BYTES ;
-			return true ;
-		}
-		bool deserialise(const void *data,uint32_t pktsize,uint32_t& offset)
-		{
-			if(offset + SIZE_IN_BYTES > pktsize)
-				return false ;
+	inline static uint32_t serial_size() { return SIZE_IN_BYTES; }
 
-			memcpy(bytes,&((uint8_t*)data)[offset],SIZE_IN_BYTES) ;
-			offset += SIZE_IN_BYTES ;
-			return true ;
-		}
-	private:
-		std::string toStdString(bool upper_case) const ;
+	bool serialise(void* data,uint32_t pktsize,uint32_t& offset) const
+	{
+		if(offset + SIZE_IN_BYTES > pktsize) return false;
+		memmove( &(reinterpret_cast<uint8_t*>(data))[offset],
```

Comment:
no need for memmove here. If the areas overlap, then something's already wrong in the code!